File: data_structures_and_algorithms/graphs/dijkstra/python/dijkstra.py
#!/bin/python3
import unittest

"""
Sample Graph:


      ---> (A) --- 1
   6 --     ^    ----
    --      |        v
(start)    3|      (end)
    --      |        ^
   2 --     |    ----
      ---> (B) --- 5

The shortest route in this graph is 6 units long:

    start --> B --> A --> end
           2  +  3  +  1  = 6

"""

sample_graph = dict()
sample_graph["start"] = { "a": 6, "b": 2 }
sample_graph["a"] = { "end": 1 }
sample_graph["b"] = { "a": 3, "end": 5 }
sample_graph["end"] = {}

### Table Breakdown

### Initial
#   node    cost    processed   parent
#   ---     ---     ---         ---
#   start   0       False       None
#   A       None    False       None
#   B       None    False       None
#   end     None    False       None

### Processed Start
#   node    cost    processed   parent
#   ---     ---     ---         ---
#   start   0       True        None
#   A       6       False       Start
#   B       2       False       Start
#   end     None    False       None

### Processed B
#   node    cost    processed   parent
#   ---     ---     ---         ---
#   start   0       True        None
#   A       5       False       B
#   B       2       True        Start
#   end     7       False       B

### Processed A
#   node    cost    processed   parent
#   ---     ---     ---         ---
#   start   0       True        None
#   A       5       True        B
#   B       2       True        Start
#   end     6       False       A

# Then build the path backwards from "end" by checking "parent"
# end -> A -> B -> start

# Reverse the list to get the final path
# start -> B -> A -> end






# Update the table with the new costs of neighbor nodes
# Mark this node as processed
def update_table(parent, graph, table):
    for key in graph[parent].keys():
        # New Cost = (cost_to_parent) + (cost_from_parent_to_node)
        new_cost = table[parent]["cost"] + graph[parent][key]
        old_cost = table[key]["cost"]

        if table[key]["cost"] is None or new_cost < old_cost:
            table[key]["cost"] = new_cost
            table[key]["parent"] = parent

    table[parent]["processed"] = True

# Find the next cheapest node
def get_next_cheapest(table):
    cheapest_node = None
    for key in table.keys():
        if not table[key]["processed"]:
            if cheapest_node is None or table[key]["cost"] < table[cheapest_node]["cost"]:
                cheapest_node = key
    return cheapest_node

# Build the final path backward by recursively adding parents
def get_final_path(table, end):
    final_path= []
    node = end
    while node is not None:
        final_path.append(node)
        node = table[node]["parent"]
    final_path.reverse()
    return final_path

def dijkstra(graph, start, end):
    # Initialize the table
    # Table entries contain nodes, costs, a flag for marking processed nodes, and the parent node
    table = {}
    for key in graph.keys():
        table[key] = {"cost": None, "processed": False, "parent": None }

    # We start at the first node for free, so initialize it with 0 cost
    table[start]["cost"] = 0

    # Once the next-cheapest node is the target, we know that there's no faster path than the one we've got right now.
    table[end]["processed"] = True

    current_node = start
    while current_node is not None:
        update_table(current_node, graph, table)
        current_node = get_next_cheapest(table)

    return get_final_path(table, end)

# Unit Tests
class DijkstraTest(unittest.TestCase):

    def test_basic(self):
        self.assertEqual(dijkstra(sample_graph, "start", "end"), ['start', 'b', 'a', 'end'])

File: docs/data_structures_and_algorithms/graphs/dijkstra/cpp/dijkstra.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <utility>
#include <limits>
#include <list>

using namespace std;

class TableEntry {
    public:
        TableEntry() {}
        TableEntry(string name, int cost, bool isProcessed, string parent): name(name), cost(cost), isProcessed(isProcessed), parent(parent) {}
        string name;
        int cost;
        bool isProcessed;
        string parent;
};

using my_graph_t = unordered_map<string,unordered_map<string,int>>;
using my_table_t = unordered_map<string, TableEntry>;

// Table:
// | NodeName | Cost | isProcessed? | Parent Node |


void print_table(unordered_map<string, TableEntry> &node_table) {
    cout << "\n\n";
    for (auto entry : node_table) {
        cout <<
            entry.first                                     << " | " <<
            entry.second.cost                               << " | " <<
            (entry.second.isProcessed ? "True" : "False")   << " | " << 
            entry.second.parent                             << endl; 
    }
    cout << "\n\n";
}


void process_node(string current_node, my_graph_t &graph, my_table_t &node_table) {

    cout << "Processing node: " << current_node << endl;
    // Update neighbors
    for (auto &neighbor : graph[current_node]) {

        // If this neighbor isn't present in the table, it's the first time we've seen it
        if (node_table.find(neighbor.first) == node_table.end()) {
            cout << "first time seeing " << neighbor.first << endl;
            node_table.emplace(neighbor.first, TableEntry(neighbor.first, numeric_limits<int>::max(), false, current_node));
        }

        int new_cost = node_table[current_node].cost + graph[current_node][neighbor.first];
        int old_cost = node_table[neighbor.first].cost;

        if (new_cost < old_cost) {
            node_table[neighbor.first].cost = new_cost;
            node_table[neighbor.first].parent = current_node;
        }
    }

    // Mark this node as processed
    node_table[current_node].isProcessed = true;
}

string get_next_cheapest(my_table_t &table) {
    TableEntry *current_cheapest = nullptr;

    for (auto &row : table) {

        // Find a node that:
        // - hasn't been processed
        // - cheaper than the current best
        if (row.second.isProcessed) {
            continue;
        } else if (!current_cheapest || (row.second.cost < current_cheapest->cost)) {
            current_cheapest = &row.second;
        }
    }

    if (current_cheapest) {
        return current_cheapest->name;
    } else {
        return "";
    }
}

// Build the final path backwards by repeatedly jumping to parent
void get_final_path(my_table_t &table, string start, string end) {
    list<string> final_path;
    string current_node = end;
    while (current_node != "") {
        final_path.push_front(current_node);
        current_node = table[current_node].parent;
    }
    for (auto item: final_path) {
        cout << item << endl;
    }
}

// Find the cheapest path from start to end
int dijkstra(my_graph_t &graph, string start, string end) {

    // Create the node-cost table
    my_table_t table;
    table.emplace(start, TableEntry(start, 0, false, ""));
    print_table(table);

    string current_node = get_next_cheapest(table);
    while (current_node != "" && current_node != end) {

        // Process the current node
        process_node(current_node, graph, table);
        print_table(table);

        // Get the next-cheapest node
        current_node = get_next_cheapest(table);
        print_table(table);
    }
    
    // We've found a path, now unfurl it
    if (current_node == end) {
        get_final_path(table, start, end);
    }

    return 0;
}


int main() {
    cout << "Hello World!" << endl;

    unordered_map<string, unordered_map<string,int>>sample_graph({
        {"start", {
            {"a", 6},
            {"b", 2}
        }},

        {"a", {
            {"end", 1}
        }},

        {"b", {
            {"a", 3},
            {"end", 5}
        }},

        {"end", {
        }}
    });

    dijkstra(sample_graph, "start", "end");

    return 0;
}
